### PR TITLE
Subbu/goal generation

### PR DESCRIPTION
**Pull request by SubramanianKrish**
Realsense namespace corrected in sim
static map params yaml removed and moved to gloobal params file in both sim and hw
depth registered on rgb topic
goal generation package created (couple of nodes, couple of launch files)
Done upto projecting lane lines in pcl version onto the costmap using movebase obstacle layer plugin

### CODE REVIEW

File: autovalet_gazebo/models/cmu_ground_plane/materials/scripts/cmu_ground_plane.material [hunk 1]
```diff
@@ -10,7 +10,7 @@ material CMUGroundPlane/Image
 
             texture_unit
             {
-                texture dark-garage_onlyRightCurve.jpg
+                texture garage_green.jpg
```
Comment (kob51) [inline autovalet_gazebo/models/cmu_ground_plane/materials/scripts/cmu_ground_plane.material]: don't push this change to master.

File: autovalet_goal_generation/launch/seg_map.launch [hunk 2]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0"?>
```
Comment (kob51) [inline autovalet_goal_generation/launch/seg_map.launch]: is this file not a lot of copy/pasted from parking_lot.launch? could we clean that up?

File: autovalet_goal_generation/scripts/transfer_lane.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    # Read camera info by looking up only one message
+    self.camera = rospy.wait_for_message(cameraInfo_topic, CameraInfo)
+    self.lane_cloud_pub = rospy.Publisher(laneCloud_topic, PointCloud2, queue_size=1)
+
+    # Looking to synchronize both the topics within a 1/10th of a second
+    self.ats = ApproximateTimeSynchronizer([self.depth_sub, self.segmap_sub], queue_size=2, slop=0.1)
+    self.ats.registerCallback(self.callback)
+
+    # Define the camera matrix individual values for projection
+    self.fx     = self.camera.K[0]
```
Comment (kob51) [inline autovalet_goal_generation/scripts/transfer_lane.py:43]: where's the 5th dof of K?

File: autovalet_goal_generation/scripts/transfer_lane.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    # Close the segmap to remove small holes in the segmented lane
+    segmap_processed = cv2.morphologyEx(segmap_image, cv2.MORPH_CLOSE, self.dilation_kernel)
+    # Generate new mask with lane boundary
+    lane_lines = cv2.Canny(segmap_processed, 100, 200)
+    # Perform morphological opening to increase width of lane line
+    lane_lines = cv2.dilate(lane_lines,self.dilation_kernel,iterations = 1)
+
+    # Overlay mask
+    lane_depth_image = cv2.bitwise_and(depth_image, depth_image, mask = lane_lines/255)
```
Comment (kob51) [inline autovalet_goal_generation/scripts/transfer_lane.py:64]: is "depth_image" both src and destination? wondering why it comes up twice in parameters

File: autovalet_goal_generation/scripts/transfer_lane.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+	[height, width] = depth_image.shape
+
+	nx = np.linspace(0, width-1, width)
+	ny = np.linspace(0, height-1, height)
+	u, v = np.meshgrid(nx, ny)
+	x = (u.flatten() - self.cx)/self.fx
+	y = (v.flatten() - self.cy)/self.fy
+
+	z = depth_image.flatten() / 1000.0;
```
Comment (kob51) [inline autovalet_goal_generation/scripts/transfer_lane.py:104]: so you normalize Z then multiply X and Y by Z?

File: autovalet_navigation/config/husky/static_map_params.yaml [hunk 6]
```diff
@@ -1,2 +0,0 @@
-map_topic: "/rtabmap/grid_map"
```
Comment (kob51) [inline autovalet_navigation/config/husky/static_map_params.yaml:1]: why did you delete these files (same one in husky and sim dirs)? where does this info live now?

File: autovalet_navigation/config/sim/global_costmap_params.yaml [hunk 7]
```diff
@@ -1,13 +1,24 @@
+# Added lane as a plugin for obstacle layer (?)
 global_costmap:
   plugins:
       - {name: static_layer,   type: "costmap_2d::StaticLayer"}
       - {name: inflation,      type: "costmap_2d::InflationLayer"}
+      - {name: lane,           type: "costmap_2d::ObstacleLayer"}
+
+  lane:
```
Comment (kob51) [inline autovalet_navigation/config/sim/global_costmap_params.yaml:8]: why does this not live in the costmap_common_params.yaml file next to the other pcl sources?

File: autovalet_navigation/config/sim/global_costmap_params.yaml [hunk 8]
```diff
@@ -1,13 +1,24 @@
...
-  publish_frequency: 1.0
+  publish_frequency: 10.0
   static_map: false
   rolling_window: true
   always_send_full_costmap: true
-  map_type: costmap
\ No newline at end of file
+  map_type: costmap
+
+  static_layer:
```
Comment (kob51) [inline autovalet_navigation/config/sim/global_costmap_params.yaml:22]: oh you moved it here. make sure you do the same thing on the `husky` version of this file so we don't forget about it.

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 3 above)
Comment (evamo0508) [reply autovalet_goal_generation/scripts/transfer_lane.py:43]: I believe there are only 4: cx, cy, fx, fy

**Review by mishra-nikita — DISMISSED** [review]
kewlls

File: autovalet_gazebo/launch/valet_rtab_sim.launch [hunk 9]
```diff
@@ -47,7 +47,7 @@
         <!-- ICP parameters -->
         <param name="Icp/PointToPlane"        type="string" value="true"/>
         <param name="Icp/Iterations"          type="string" value="20"/>
-        <param name="Icp/VoxelSize"           type="string" value="0.01"/> <!-- 0.01 works pretty well -->
+        <param name="Icp/VoxelSize"           type="string" value="0.1"/> <!-- 0.01 works pretty well -->
```
Comment (evamo0508) [inline autovalet_gazebo/launch/valet_rtab_sim.launch:50]: why is this? Shouldn't increasing voxel size lower the localization accuracy? or did we want to prevent some errors happening due to small voxelSize?

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 5 above)
Comment (evamo0508) [reply autovalet_goal_generation/scripts/transfer_lane.py:104]: the default z values come in `mm`, but we want meters, that's why we divided them by 1000.0

**Review by evamo0508 — COMMENTED** [review]
Why are there two `generate_gt_segmap.py`?

File: autovalet_goal_generation/launch/seg_map.launch (hunk 2 above)
Comment (SubramanianKrish) [reply autovalet_goal_generation/launch/seg_map.launch]: Yeah. it's mostly a copy of parking_lot launch. How do you want me to clean it up?

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 3 above)
Comment (SubramanianKrish) [reply autovalet_goal_generation/scripts/transfer_lane.py:43]: skew is assumed zero

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 4 above)
Comment (SubramanianKrish) [reply autovalet_goal_generation/scripts/transfer_lane.py:64]: I used this as a reference .. didn't dig too deep there
https://docs.opencv.org/master/d0/d86/tutorial_py_image_arithmetics.html

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 5 above)
Comment (SubramanianKrish) [reply autovalet_goal_generation/scripts/transfer_lane.py:104]: The plugin gives the output in mm. Similar to actual camera if i'm not wrong

File: autovalet_navigation/config/husky/static_map_params.yaml (hunk 6 above)
Comment (SubramanianKrish) [reply autovalet_navigation/config/husky/static_map_params.yaml:1]: file cleanup. Neater that way

File: autovalet_navigation/config/sim/global_costmap_params.yaml (hunk 7 above)
Comment (SubramanianKrish) [reply autovalet_navigation/config/sim/global_costmap_params.yaml:8]: It didn't work if I put it in costmap common. Had to be here for it to work

File: autovalet_navigation/config/sim/global_costmap_params.yaml (hunk 8 above)
Comment (SubramanianKrish) [reply autovalet_navigation/config/sim/global_costmap_params.yaml:22]: Yup. Done.

File: autovalet_gazebo/launch/valet_rtab_sim.launch (hunk 9 above)
Comment (SubramanianKrish) [reply autovalet_gazebo/launch/valet_rtab_sim.launch:50]: Oh that's to prevent all the error messages about leaf size coming up on the terminal. Sure smaller voxels are more accurate but we thought they may be eating up compute at the planner stage. This was one of the experimentation changes. Don't know how it made through to the PR but I think we can keep it as it removes the warning messages on the slam terminal

**Comment by SubramanianKrish** [comment]
> Why are there two `generate_gt_segmap.py`?

I've deleted the ghost file created under lane_detection. It should not be in the PR anymore

File: autovalet_goal_generation/launch/seg_map.launch [hunk 10]
```diff
@@ -0,0 +1,85 @@
+<!--
```
Comment (kob51) [inline autovalet_goal_generation/launch/seg_map.launch:1]: so it seems like a lot of the calls in this file could be separated out into its own file. and thne maybe both parking_lot.launch and this file could call that file? idk. could be more trouble when it's worth, i just don't love the idea of copy-pasting code. your call

File: autovalet_goal_generation/scripts/transfer_lane.py (hunk 5 above)
Comment (kob51) [reply autovalet_goal_generation/scripts/transfer_lane.py:104]: i get the metric conversion, now look at lines 105 and 106. multiplying x and y by z

Comment (evamo0508) [reply autovalet_goal_generation/scripts/transfer_lane.py:104]: that's a part of the camera coordinate transformation stuff which @SubramanianKrish  and I had gone through the derivations the other day. just think of the `s[u v 1] = KD[x y z 1]` equation and mess around with the three separate equations for each element, you'll get it.

File: autovalet_goal_generation/launch/seg_map.launch (hunk 10 above)
Comment (SubramanianKrish) [reply autovalet_goal_generation/launch/seg_map.launch:1]: Done. Check it out now!

**Review by kob51 — APPROVED** [review]

**Review by evamo0508 — APPROVED** [review]
